vcd/resource_vcd_vm_affinity_rule_test.go

```diff
@@ -0,0 +1,607 @@
+			return fmt.Errorf("error: could not find VDC: %s", err)
+		}
+
+		vmAffinityRule, err := vdc.GetVmAffinityRuleById(rs.Primary.ID)
+		if err != nil {
+			return fmt.Errorf("error: could not find VM affinity rule: %s", err)
+		}
+
+		*rule = *vmAffinityRule
+
+		return nil
+	}
+}
+
+// testAccCheckVmAffinityRuleDestroy checks that the affinity rule was destroyed
+// The 'rule' object was created when checking for creation
+func testAccCheckVmAffinityRuleDestroy(rule *govcd.VmAffinityRule, orgName, vdcName string) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+
+		if rule == nil || rule.VmAffinityRule == nil {
+			return fmt.Errorf("affinity rule passed to destroy check is null")
+		}
+		conn := testAccProvider.Meta().(*VCDClient)
+		org, err := conn.VCDClient.GetAdminOrgByName(orgName)
+		if err != nil {
+			return fmt.Errorf("error: could not find Org: %s", err)
+		}
+		vdc, err := org.GetVDCByName(vdcName, false)
+		if err != nil {
+			return fmt.Errorf("error: could not find VDC: %s", err)
+		}
+
+		var vmAffinityRule *govcd.VmAffinityRule
+		for N := 0; N < 15; N++ {
+
+			vmAffinityRule, err = vdc.GetVmAffinityRuleById(rule.VmAffinityRule.ID)
+			if err != nil && vmAffinityRule == nil {
+				break
+			}
+			time.Sleep(time.Second)
+		}
+
+		if err != govcd.ErrorEntityNotFound {
+			return fmt.Errorf("VM affinity rule %s was not destroyed", rule.VmAffinityRule.Name)
+		}
+		if vmAffinityRule != nil {
+			return fmt.Errorf("rule %s was found", rule.VmAffinityRule.Name)
+		}
+		return nil
+	}
+}
+
+// importStateIdByAffinityRule runs the import of a VM affinity rule using the resource ID
+func importStateIdByAffinityRule(resource string) resource.ImportStateIdFunc {
+	return func(s *terraform.State) (string, error) {
+		rs, ok := s.RootModule().Resources[resource]
+		if !ok {
+			return "", fmt.Errorf("resource not found: %s", resource)
+		}
+
+		if rs.Primary.ID == "" {
+			return "", fmt.Errorf("no ID is set for %s resource", resource)
+		}
+
+		importId := testConfig.VCD.Org + "." + testConfig.VCD.Vdc + "." + rs.Primary.ID
+		if testConfig.VCD.Org == "" || testConfig.VCD.Vdc == "" || rs.Primary.ID == "" {
+			return "", fmt.Errorf("missing information to generate import path: %s", importId)
+		}
+		return importId, nil
+	}
+}
+
+// makeEmptyVapp creates a given vApp without any VM
+func makeEmptyVapp(vdc *govcd.Vdc, name string) (*govcd.VApp, error) {
+
+	err := vdc.ComposeRawVApp(name)
+	if err != nil {
+		return nil, err
+	}
+	vapp, err := vdc.GetVAppByName(name, true)
+	if err != nil {
+		return nil, err
+	}
+	initialVappStatus, err := vapp.GetStatus()
+	if err != nil {
+		return nil, err
+	}
+	if initialVappStatus != "RESOLVED" {
+		err = vapp.BlockWhileStatus(initialVappStatus, testConfig.Provider.MaxRetryTimeout)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return vapp, nil
+}
+
+// makeEmptyVm creates an empty VM inside a given vApp
+func makeEmptyVm(vapp *govcd.VApp, name string) (*govcd.VM, error) {
```

Comment:
Didn't we have a function to create empty VMs already? `addEmptyVm`?

Reply:
That is what `makeEmptyVm` is using